services/github/github-tag.service.js

```diff
@@ -73,14 +91,26 @@ class GithubTag extends GithubAuthV4Service {
     }
   }
 
-  async fetch({ user, repo, sort }) {
-    const limit = sort === 'semver' ? 100 : 1
+  async fetch({ user, repo, sort, prefix, subpackage }) {
```

Comment:
I'm confused about what we're trying to achieve here. There's 2 variables here: `prefix` and `subpackage` but `prefix` doesn't seem to actually get used and 'subpackage' is an odd choice of concept/variable name for a generic github badge. You've talked about go subpackages, but this is a generic github tag badge. We shouldn't be introducing a concept which is specific to some particular language ecosystem here.

It seems like we just want a single generic "prefix" filter var, which may or may not end with a / and could be a go subpackage, or any other prefix. If you want to use `refPrefix` where possible (due to the 100 objects per page limit), maybe we just check for `prefix.endsWith('/')` ?

Reply:
The tags returned by `refPrefix`do not include the prefix so if you had a tag `example/v0.0.1` the subpackage route returns `v0.0.1` and the prefix would return `example/v0.0.1` Since those are different I didn't want to one param to auto switch between them because it would make for a bad experience.

`prefix` here is just used to get 100 tags instead of just the newest which is the same experience as `sort === 'semver'`. Filtering by prefix happens in the function `getLatestTag `.

I agree that you don't want language specific things so do you have a better name for a prefix that end with `/`? Maybe `folder` or `pathPrefix`

Reply:
OK, and is the reason we are talking about a variant that strips the prefix because we're looking for a way to use `refPrefix`, or because that was the original intent? As I said earlier, I think suggesting the use of `refPrefix` was not super-useful on my part and may have sent us down the wrong road - I hadn't realised the limitations when I suggested it.

I think if we do want both a variant that includes the prefix and a variant that excludes it (because that is actually useful), then the way to express this would be with a flag param called `exclude_prefix` or maybe `strip_prefix` which works regardless of whether `prefix` ends with a `/` or not.

i.e: if your list of tags is

https://github.com/badges/shields/tags

- `?prefix=server-` ==> latest tag is `server-2022-08-01` (this should be the default, because that is the actual tag name)
- `?prefix=server-&exclude_prefix` ==> latest tag (excluding prefix) is `2022-08-01`

We express these flags in the `queryParamSchema` using `exclude_prefix: Joi.equal('')`.

How does that sound?

Reply:
I can remove the `subpackage` stuff if you want but I actually like the two separate flows. In my use case for Go the `subpackage` flow is the better fix. 

Since the subpackage flow uses `refPrefix` it doesn't have the issue of needing to be within the last 100 tags. Also in the case that you want to filter the subpackage tags by a `prefix` that would work as well. 

I personally don't think excluding the prefix is often useful but `refPrefix` does that automatically. I don't want to have some code checking for a prefix and adding or removing it depending on if the prefix ends in `/` or not. That just seems too complicated. I just think the subpackage flow needs a better name.

Reply:
At a minimum `subpackage` is going to have to be renamed as the user-facing query parameter. It's crucial that we separate inner details, like behavior of a GraphQL query parameter, from the user interface surface. The parameters, and their respective names, need to clearly indicate to any and every user how that parameter will behave and reflect in the final badge.

It shouldn't be driven by one particular usage pattern (e.g. `subpackage` may make sense for tags from a repo that's Go-based, but would be puzzling in say a JavaScript repository).

At a high level, I feel like the end state in terms of badge scenarios is pretty straightforward, wherein there's a singular repository that's a monorepo of sorts or which at least contains multiple components that are versioned and tagged independently, and folks in those scenarios want to be able to get github repo tag-based version badges for those respective components.

`prefix` makes sense as a query param in this context, as we're allowing users to filter down to a subset of the tags. `subpackage` as the name does not immediately scream "the selected tag version value based on the prefix, but with the prefix stripped".

We can do whatever we need to internally, and certainly still support both having the message be prefix-ful and prefix-less to support different user needs, but we just need to find a way to surface up those two separate cases in a more intuitive manner.

I'd echo Chris' prior suggestion for how to handle this because it's very explicit and clear. Yes, that would require 2 query params instead of just 1 in the badge url, but always important to remember that badge urls are copied/written once and then never looked at again (not something regularly typed by a human). If you'd still prefer trying to support both paths with a single query parameter, then we'd have to rename both params to make each explicit (e.g. `filterAndStripPrefix`)

Reply:
Thanks for that feedback! I think this final change makes everyone happy.

I removed `subpackage` and added a bool of `trimPrefixAtSlash` that splits by the last slash into a prefix used in a filter and a prefix used in graphql. When it is not checked the whole prefix is used in the filter.